sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc

```diff
@@ -0,0 +1,555 @@
+  template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+  using value_t = property_value<work_group_progress_key, std::integral_constant<forward_progress_guarantee, Guarantee>, std::integral_constant<execution_scope, CoordinationScope>>;
+};
+
+struct sub_group_progress_key {
+  template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+  using value_t = property_value<sub_group_progress_key, std::integral_constant<forward_progress_guarantee, Guarantee>, std::integral_constant<execution_scope, CoordinationScope>>;
+};
+
+struct work_item_progress_key {
+  template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+  using value_t = property_value<work_item_progress_key, std::integral_constant<forward_progress_guarantee, Guarantee>, std::integral_constant<execution_scope, CoordinationScope>>;
+};
+
+template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+inline constexpr work_group_progress_key::value_t<Guarantee, CoordinationScope> work_group_progress;
+
+template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+inline constexpr sub_group_progress_key::value_t<Guarantee, CoordinationScope> sub_group_progress;
+
+template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+inline constexpr work_item_progress_key::value_t<Guarantee, CoordinationScope> work_item_progress;
+
+template <> struct is_property_key<work_group_progress_key> : std::true_type {};
+template <> struct is_property_key<sub_group_progress_key> : std::true_type {};
+template <> struct is_property_key<work_item_progress_key> : std::true_type {};
+
+}
+----
+
+|===
+|Property|Description
+
+|`work_group_progress`
+|The `work_group_progress` property adds the requirement that the kernel must be
+ launched with the specified forward progress guarantees for threads of execution
+ created at `execution_scope::work_group` scope, with the specified
+ coordination scope.
+
+|`sub_group_progress`
+|The `sub_group_progress` property adds the requirement that the kernel must be
+ launched with the specified forward progress guarantees for threads of execution
+ created at `execution_scope::sub_group` scope, with the specified coordination
+ scope.
+
+|`work_item_progress`
+|The `work_item_progress` property adds the requirement that the kernel must be
+ launched with the specified forward progress guarantees for threads of execution
+ created at `execution_scope::work_item` scope, with the specified coordination
+ scope.
+
+|===
+
+
+=== Launch queries [[launch]]
+
+An implementation's ability to satisify requests for specific forward progress
+guarantees for a kernel may be dependent upon: the kernel itself, and the
+features that it uses; the queue (and underlying device) to which the kernel is
+submitted; and the kernel's launch configuration.
+
+It is a user's responsibility to ensure that a kernel requesting specific
```

Comment:
It seems that the intent is that these queries return values that are dependent upon the kernel properties like `work_group_progress`, etc.  This can only work if those properties are compile-time constants and if the property list itself is a compile-time constant.  However, it seems like kernel property lists are not defined this way.  For example, what would prevent someone from writing code like:

```
properties props1{/*...*/};
properties props2{/*...*/};
cgh.parallel_for(rng, ((cond) ? props1 : props2), [=](int i) {/*...*/});
```

Did we expect that the set of kernel properties is always known statically?  Perhaps this is a flaw in our kernel properties extension.

Reply:
The type of a `properties` variable reflects (in its template arguments) all of the properties it contains (including run-time properties).  So `((cond) ? props1 : props2)` is only valid C++ if `props1` and `props2` differ only in the values of their runtime values -- otherwise it's equivalent to something like `((cond) ? std::vector<int>{} : std::array<int, 1>{})`.

Any property that might impact the way that a kernel gets compiled must be a compile-time constant property, because there's no guarantee that SYCL compilers will provide a JIT mechanism.  All the kernel properties we've specified so far (i.e., progress guarantees, `use_root_sync`, `sub_group_size`, etc) are compile-time constant properties.

You're right that we need to be careful here, but I think we've been careful so far.  We would run into problems if somebody introduced a runtime property intended to impact the return values of these launch queries, but I don't think we should introduce properties like that.

Reply:
Good point!  In fact, we can determine the kernel properties solely by looking at the type of the `PropertyList` template parameter.  Few!  I'm glad we do not have a flaw with the property list design.  Similar logic applies to the "embedded" kernel properties:

```
struct KernelFunctor {
  auto get(sycl::ext::oneapi::experimental::properties_tag) {
    return sycl::ext::oneapi::experimental::properties{/*...*/};
  }
};
```

We can determine the properties solely from the return type of the `get` function, and it would be a syntax error if the user had conditional logic that returned different property lists (with different types) depending on the condition.